src/architecture/riscv.rs

```diff
@@ -23,56 +24,188 @@ impl CPURegister {
+pub enum PTE32Size {
+    /// 4 bytes
+    #[default]
+    PTP32,
+    /// 4KB, 1024 * 4 bytes
+    PTE4KB,
+    /// 4MB, 1024 * 1024 * 4 bytes
+    PTE4MB,
+}
+
+/// Represents a RISC-V SV39 and SV48 page table entry size.
+#[derive(Debug, Clone, Serialize, Deserialize, Hash, Eq, PartialEq, Ord, PartialOrd, Default)]
+pub enum PTE64Size {
+    /// 8 bytes
+    #[default]
+    PTP64,
+    /// 4KB, 512 * 8 bytes
+    PTE4KB,
+    /// 2MB, 256 * 8 bytes
+    PTE2MB,
+    /// 1GB, 128 * 8 bytes
+    PTE1GB,
+    /// 512GB, 64 * 8 bytes
+    /// This is only available in Sv48
+    PTE512GB,
+}
+
+/// Represents a RISC-V SV32 page table entry.
 /// It holds the mapping between a virtual address of a page and the address of a physical frame.
 /// There is also auxiliary information about the page such as a present bit, a dirty or modified bit,
 /// address space or process ID information, amongst others.
-#[derive(Debug, Clone, Serialize, Deserialize, Hash, Eq, PartialEq, Ord, PartialOrd)]
-pub struct PageTableEntry {
-    pub address: u64,
-    pub flags: u64,
+#[derive(Debug, Clone, Serialize, Deserialize, Hash, Eq, PartialEq, Ord, PartialOrd, Default)]
+pub struct PageTableEntry<A, F, S> {
+    pub address: A,
+    pub flags: F,
+    pub size: S,
 }
 
-impl PageTableEntry {
-    pub fn new(address: u64, flags: u64) -> Self {
-        Self { address, flags }
+impl PageTableEntry<u32, PTE32Flags, PTE32Size> {
+    pub fn new(address: u32, flags: u32, size: PTE32Size) -> Self {
+        let flags = PTE32Flags(flags);
+        Self {
+            address,
+            flags,
+            size,
+        }
     }
 
     pub fn is_supervisor(&self) -> bool {
-        todo!()
+        self.flags.supervisor()
     }
 }
 
-impl PageTableEntryTrait for PageTableEntry {
+impl PageTableEntryTrait for PageTableEntry<u32, PTE32Flags, PTE32Size> {
+    type Address = u32;
+    type Flags = PTE32Flags;
+    type Size = PTE32Size;
+
+    fn is_dirty(&self) -> bool {
+        self.flags.dirty()
+    }
+
+    fn is_accessed(&self) -> bool {
+        self.flags.accessed()
+    }
+
+    fn is_global(&self) -> bool {
+        self.flags.global()
+    }
+
+    fn is_readable(&self) -> bool {
+        self.flags.read()
+    }
+
+    fn is_writable(&self) -> bool {
+        self.flags.write()
+    }
+
+    fn is_executable(&self) -> bool {
+        self.flags.exec()
+    }
+}
+
+impl PageTableEntry<u64, PTE64Flags, PTE64Size> {
+    pub fn new(address: u64, flags: u64, size: PTE64Size) -> Self {
+        let flags = PTE64Flags(flags);
+        Self {
+            address,
+            flags,
+            size,
+        }
+    }
+
+    pub fn is_supervisor(&self) -> bool {
+        self.flags.supervisor()
+    }
+}
+
+impl PageTableEntryTrait for PageTableEntry<u64, PTE64Flags, PTE64Size> {
     type Address = u64;
-    type Flags = u64;
+    type Flags = PTE64Flags;
+    type Size = PTE64Size;
 
-    // FIXME: Implement the following methods
```

Comment:
are those methods really useful? We could just check via pte.flags.xxx instead of the methods defined here

Reply:
It will allow us to have a more streamlined API